ros2_control PDO commands (#219)
husarion/panther_ros: panther_hardware_interfaces/README.md

Reviewer: something I have missed before - according to our README writing guide default values should be bold. Eg.

- `param` [*type*, default: **value**]: description.

 - `max_rpm_motor_speed` [*float*, default: **3600.0**]: max RPM speed set in the Roboteq driver (MXRPM parameter)
 - `gearbox_efficiency` [*float*, default: **0.75**]: measured efficiency, used for converting read current to torque, can vary depending on different factors such as temperature and wear

CAN settings
 - `can_interface_name` [*string*, default: **panther_can**]: name of the CAN interface
 - `master_can_id` [*int*, default: **3**]: CAN ID of the master device (set as in [canopen_configuration.yaml](./config/canopen_configuration.yaml))
 - `front_driver_can_id` [*int*, default: **1**]: CAN ID defined in the properties of Roboteq (set as in [canopen_configuration.yaml](./config/canopen_configuration.yaml))
 - `rear_driver_can_id` [*int*, default: **2**]: CAN ID defined in the properties of Roboteq (set as in [canopen_configuration.yaml](./config/canopen_configuration.yaml))
 - `sdo_operation_timeout_ms` [*int*, default: **100**]: timeout of the SDO operations, currently no SDO operation is required in RT operation, so this timeout can be set to a higher value

>  - `pdo_motor_states_timeout_ms` [*int*, default: **15**]: depends on the frequency at which Roboteq is configured to send motor states (PDO 1 and 2) data. By default there should 10 **[ms]** between received data, if it takes more than `pdo_motor_states_timeout_ms`, a motor states read error is triggered. Default value is set to be expected period +50% margin.

Reviewer: ```suggestion
 - `pdo_motor_states_timeout_ms` [*int*, default: **15**]: depends on the frequency at which Roboteq is configured to send motor states (PDO 1 and 2) data. By default there should be 10 **[ms]** between received data, if it takes more than `pdo_motor_states_timeout_ms`, a motor states read error is triggered. Default value is set to be expected period +50% margin.
```

>  - `pdo_driver_state_timeout_ms` [*int*, default: **75**]: depends on the frequency at which Roboteq is configured to send driver state (PDO 3 and 4) data. By default there should 50 **[ms]** between received data, if it takes more than `pdo_driver_state_timeout_ms`, a driver state read error is triggered. Default value is set to be expected period +50% margin.

Reviewer: ```suggestion
 - `pdo_driver_state_timeout_ms` [*int*, default: **75**]: depends on the frequency at which Roboteq is configured to send driver state (PDO 3 and 4) data. By default there should be 50 **[ms]** between received data, if it takes more than `pdo_driver_state_timeout_ms`, a driver state read error is triggered. Default value is set to be expected period +50% margin.
```

 - `max_roboteq_initialization_attempts` [*int*, default: **5**]: in some cases, an SDO error can happen during initialization, it is possible to configure more attempts, before escalating to general error
 - `max_roboteq_activation_attempts` [*int*, default: **5**]: similar to initialization, it is possible to allow some SDO errors before escalating to error
 <!-- TODO: with GPIODriver it should no longer be needed -->
 - `max_safety_stop_attempts` [*int*, default: **20**]: how many attempts to activate safety stop will be taken before failing
 - `max_write_pdo_cmds_errors_count` [*int*, default: **4**]: how many consecutive errors can happen before escalating to general error
 - `max_read_pdo_motor_states_errors_count` [*int*, default: **4**]: how many consecutive errors can happen before escalating to general error
 - `max_read_pdo_driver_state_errors_count` [*int*, default: **20**]: how many consecutive errors can happen before escalating to general error

> 

Reviewer 2: ```suggestion
 - `can_interface_name` [*string*, default: **panther_can**]: name of the CAN interface
 - `master_can_id` [*int*, default: **3**]: CAN ID of the master device (set as in [canopen_configuration.yaml](./config/canopen_configuration.yaml))
 - `front_driver_can_id` [*int*, default: **1**]: CAN ID defined in the properties of Roboteq (set as in [canopen_configuration.yaml](./config/canopen_configuration.yaml))
 - `rear_driver_can_id` [*int*, default: **2**]: CAN ID defined in the properties of Roboteq (set as in [canopen_configuration.yaml](./config/canopen_configuration.yaml))
 - `sdo_operation_timeout_ms` [*int*, default: **100**]: timeout of the SDO operations, currently no SDO operation is required in RT operation so that this timeout can be set to a higher value
 - `pdo_motor_states_timeout_ms` [*int*, default: **15**]: depends on the frequency at which Roboteq is configured to send motor states (PDO 1 and 2) data. By default, there should be 10 **[ms]** between received data, if it takes more than `pdo_motor_states_timeout_ms`, a motor states read error is triggered. The default value is set to be expected period +50% margin.
 - `pdo_driver_state_timeout_ms` [*int*, default: **75**]: depends on the frequency at which Roboteq is configured to send driver state (PDO 3 and 4) data. By default, there should be 50 **[ms]** between received data, if it takes more than `pdo_driver_state_timeout_ms`, a driver state read error is triggered. The default value is set to be expected period +50% margin.
 - `driver_states_update_frequency` [*float*, default: **20.0**]: as by default, the driver state is published with lower frequency, it also shouldn't be updated with every controller loop iteration. The exact frequency at which driver state is published won't match this value - it will also depend on the frequency of the controller (the exact value of the period can be calculated with the following formula `controller_frequency / ceil(controller_frequency / driver_states_update_frequency)`)
 - `max_roboteq_initialization_attempts` [*int*, default: **5**]: in some cases, an SDO error can happen during initialization, it is possible to configure more attempts, before escalating to a general error
 - `max_roboteq_activation_attempts` [*int*, default: **5**]: similar to initialization, it is possible to allow some SDO errors before escalating to error
 <!-- TODO: with GPIODriver it should no longer be needed -->
 - `max_safety_stop_attempts` [*int*, default: **20**]: how many attempts to activate safety stop will be taken before failing
 - `max_write_pdo_cmds_errors_count` [*int*, default: **4**]: how many consecutive errors can happen before escalating to general error
 - `max_read_pdo_motor_states_errors_count` [*int*, default: **4**]: how many consecutive errors can happen before escalating to general error
 - `max_read_pdo_driver_state_errors_count` [*int*, default: **20**]: how many consecutive errors can happen before escalating to general error

```

Author: I think that this change `timeout of the SDO operations, currently no SDO operation is required in RT operation so that this timeout can be set to a higher value` has slightly different meaning, and I would prefer to leave earlier version.

 - `master_can_id` [*int*, default: 3]: CAN ID of the master device (set as in [canopen_configuration.yaml](./config/canopen_configuration.yaml))
 - `front_driver_can_id` [*int*, default: 1]: CAN ID defined in the properties of Roboteq (set as in [canopen_configuration.yaml](./config/canopen_configuration.yaml))
 - `rear_driver_can_id` [*int*, default: 2]: CAN ID defined in the properties of Roboteq (set as in [canopen_configuration.yaml](./config/canopen_configuration.yaml))

>  - `sdo_operation_timeout_ms` [*int*, default: 100 [ms]]: timeout of the SDO operations, currently no SDO operation is required in RT operation, so this timeout can be set to a higher value

Reviewer: ```suggestion
 - `sdo_operation_timeout_ms` [*int*, default: 100]: timeout in **[ms]** of the SDO operations, currently no SDO operation is required in RT operation, so this timeout can be set to a higher value
```
I don't know if putting units inside is a good idea. Also, units should be bold.

Author: after adding `_ms` to the variable name I'm not sure if it necessary to add the `[ms]`

>  - `pdo_motor_states_timeout_ms` [*int*, default: 15 [ms]]: depends on the frequency at which Roboteq is configured to send motor states (PDO 1 and 2) data. By default there should 8ms between received data, if it takes more than `pdo_motor_states_timeout_ms`, a motor states read error is triggered. Default value is set to be expected period +50% margin

Reviewer: ```suggestion
 - `pdo_motor_states_timeout_ms` [*int*, default: 15]: timeout in **[ms]**, depends on the frequency at which Roboteq is configured to send motor states (PDO 1 and 2) data. By default there should be 8 **[ms]** between received data, if it takes more than `pdo_motor_states_timeout_ms`, a motor states read error is triggered. Default value is set to be expected period +50% margin
```
As above

Author: as above, not sure if `[ms]` is even needed

 - `pdo_driver_state_timeout_ms` [*int*, default: 75 [ms]]: depends on the frequency at which Roboteq is configured to send driver state (PDO 3 and 4) data. By default there should 104ms between received data, if it takes more than `pdo_driver_state_timeout_ms`, a driver state read error is triggered. Default value is set to be expected period +50% margin.

>  - `driver_states_update_frequency` [*float*, default: 20.0 [Hz]]: as by default driver state is published with lower frequency, it also shouldn't be updated with every controller loop iteration. Exact frequency at which driver state is published won't match this value - it will depend also on the frequency of the controller (exact value of period can be calculated with the following formula `controller_frequency / ceil(controller_frequency / driver_states_update_frequency)`)

Reviewer: ```suggestion
 - `pdo_driver_state_timeout_ms` [*int*, default: 75]: timeout in **[ms]**, depends on the frequency at which Roboteq is configured to send driver state (PDO 3 and 4) data. By default there should 104ms between received data, if it takes more than `pdo_driver_state_timeout_ms`, a driver state read error is triggered. Default value is set to be expected period +50% margin.
 - `driver_states_update_frequency` [*float*, default: 20.0]: as by default driver state is published with lower frequency, it also shouldn't be updated with every controller loop iteration. Exact frequency at which driver state is published won't match this value - it will depend also on the frequency of the controller (exact value of period can be calculated with the following formula `controller_frequency / ceil(controller_frequency / driver_states_update_frequency)`)
```
For frequency unit is pretty obvious

Author: here I also remove `[ms]`

 - `max_roboteq_initialization_attempts` [*int*, default: 5]: in some cases, an SDO error can happen during initialization, it is possible to configure more attempts, before escalating to general error
 - `max_roboteq_activation_attempts` [*int*, default: 5]: similar to initialization, it is possible to allow some SDO errors before escalating to error
 <!-- TODO: with GPIODriver it should no longer be needed -->
 - `max_safety_stop_attempts` [*int*, default: 20]: how many attempts to activate safety stop will be taken before failing
 - `max_write_pdo_cmds_errors_count` [*int*, default: 4]: how many consecutive errors can happen before escalating to general error
 - `max_read_pdo_motor_states_errors_count` [*int*, default: 4]: how many consecutive errors can happen before escalating to general error
 - `max_read_pdo_driver_state_errors_count` [*int*, default: 20]: how many consecutive errors can happen before escalating to general error

> [!CAUTION]

> > `max_write_pdo_cmds_errors_count`, `max_read_pdo_motor_states_errors_count`, `max_read_pdo_driver_state_errors_count`, `max_safety_stop_attempts`. `sdo_operation_timeout`, `pdo_motor_states_timeout_ms` and `pdo_driver_state_timeout_ms` are safety-critical parameters, they should be changed only in very specific cases, be sure that you know how they work and be really cautious when changing them.

Reviewer: ```suggestion
> `max_write_pdo_cmds_errors_count`, `max_read_pdo_motor_states_errors_count`, `max_read_pdo_driver_state_errors_count`, `max_safety_stop_attempts`, `sdo_operation_timeout`, `pdo_motor_states_timeout_ms` and `pdo_driver_state_timeout_ms` are safety-critical parameters, they should be changed only in very specific cases, be sure that you know how they work and be really cautious when changing them.
```
